helix-rest/src/main/java/org/apache/helix/rest/common/HelixDataAccessorWrapper.java

```diff
@@ -1,24 +1,144 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
 
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import org.apache.helix.HelixProperty;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixDataAccessor;
+import org.apache.helix.model.RESTConfig;
+import org.apache.helix.rest.client.CustomRestClient;
+import org.apache.helix.rest.client.CustomRestClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This is a wrapper for {@link ZKHelixDataAccessor} that caches the result of the batch reads it
  * performs.
  * Note that the usage of this object is valid for one REST request.
  */
 public class HelixDataAccessorWrapper extends ZKHelixDataAccessor {
+  private static final Logger LOG = LoggerFactory.getLogger(HelixDataAccessorWrapper.class);
+  private static final ExecutorService POOL = Executors.newCachedThreadPool();
+
+  static final String PARTITION_HEALTH_KEY = "PARTITION_HEALTH";
+  static final String IS_HEALTHY_KEY = "IS_HEALTHY";
+  static final String EXPIRY_KEY = "EXPIRE";
+
   private final Map<PropertyKey, HelixProperty> _propertyCache = new HashMap<>();
   private final Map<PropertyKey, List<String>> _batchNameCache = new HashMap<>();
+  protected CustomRestClient _restClient;
 
   public HelixDataAccessorWrapper(ZKHelixDataAccessor dataAccessor) {
     super(dataAccessor);
+    _restClient = CustomRestClientFactory.get();
+  }
+
+  public Map<String, Map<String, Boolean>> getAllPartitionsHealthOnLiveInstance(RESTConfig restConfig,
+                                                                                Map<String, String> customPayLoads) {
+    // Only checks the instances are online with valid reports
+    List<String> liveInstances = getChildNames(keyBuilder().liveInstances());
+    // Make a parallel batch call for getting all healthreports from ZK.
+    List<HelixProperty> zkHealthReports = getProperty(liveInstances.stream()
+        .map(instance -> keyBuilder().healthReport(instance, PARTITION_HEALTH_KEY))
+        .collect(Collectors.toList()), false);
+    Map<String, Future<Map<String, Boolean>>> parallelTasks = new HashMap<>();
+    for (int i = 0; i < liveInstances.size(); i++) {
+      String liveInstance = liveInstances.get(i);
+      Optional<ZNRecord> maybeHealthRecord = zkHealthReports != null && zkHealthReports.size() > i
```

Comment:
How does zkHealthReports.size() > i work? I am confused here. Was that the trick you did as .collect(Collectors.toList()), false)

If that's the case, can you add some comments here? It is hard to understand the code in this way.

Reply:
The **getProperty** method alone is deprecated, instead I used `getProperty(xxx, boolean throwException)`.
`List<HelixProperty> zkHealthReports` doesn't guarantee or doesn't explicitly say that it will always be of equal size of input parameters. E.g, I'd like to get 10 instances' healthReports, but the method only returns 5 objects. That's why I checked `zkHealthReports.size() > i` to avoid array index of bounds issue.
To be extra safe, 
It's possible the `zkHealthReports' is null, 'zkHealthReports' has less items than live instances parameter and each object in `zkHealthReports' is also possible null. All of these are covered in the ternary operator statement.

Reply:
If Helix does not return you complete list, this is not an efficient way. You dont know which one is missed from the array list. Then you totally mess up the result. For example, you have 6 instances and you only get 3 ZNRecords for last 3 ones. With your logic, you will fill the 3 ZNRecords as first threes health report and querying next three ones. 

Please refer the Java doc of get in BaseDataAccessor interface. getProperty uses the get() to return list of HelixProperty: https://github.com/apache/helix/blob/master/helix-core/src/main/java/org/apache/helix/BaseDataAccessor.java#L150

Reply:
'getProperty' method was not a very good design but it's already there. I may have done redundant checks to be extra safe, it won't break the functionality. If I'm asking 6 instances' records, it returns {null, record1, null, null, record, null}, it will also work.

Reply:
I've changed the line to 
```
Optional<ZNRecord> maybeHealthRecord =
          Optional.ofNullable(zkHealthReports.get(i)).map(HelixProperty::getRecord);
```
If we agree the list will always return the same number of items as request parameters